worker/src/router.rs

```diff
@@ -264,67 +271,93 @@ impl<'a, D: 'static> Router<'a, D> {
         self
     }
 
+    /// Register an HTTP handler that will catch any route that has no matched pattern for any
+    /// method. Any wildcard route will override this. Enables the use of `async/await` syntax in
+    /// the callback.
+    pub fn not_found_async<T>(mut self, func: fn(Request, RouteContext<D>) -> T) -> Self
+    where
+        T: Future<Output = Result<Response>> + 'static,
+    {
+        self.not_found_handler = Some(Handler::Async(Rc::new(move |req, route| {
+            Box::pin(func(req, route))
+        })));
+        self
+    }
+
     fn add_handler(&mut self, pattern: &str, func: Handler<'a, D>, methods: Vec<Method>) {
-        if let Ok(Match {
-            value: handler_set,
-            params: _,
-        }) = self.handlers.at_mut(pattern)
-        {
-            for method in methods {
-                handler_set[method as usize] = Some(func.clone());
-            }
-        } else {
-            let mut handler_set = [None, None, None, None, None, None, None, None, None];
-            for method in methods.clone() {
-                handler_set[method as usize] = Some(func.clone());
-            }
-            self.handlers.insert(pattern, handler_set).expect(&format!(
-                "failed to register {:?} route for {} pattern",
-                methods, pattern
-            ));
+        for method in methods {
+            self.handlers
+                .entry(method.clone())
+                .or_insert_with(Node::new)
+                .insert(pattern, func.clone())
+                .expect(&format!(
+                    "failed to register {:?} route for {} pattern",
+                    method, pattern
+                ));
```

Comment:
This allocates on each iteration of the loop. I'd suggest one of the following:
1. Move the allocation outside the loop; this would also allow you to remove `method.clone()`.
2. Use `unwrap_or_else(|| panic!(...))`.

2 is probably more efficient since `Method` is really cheap to clone.